### PR TITLE
Replace call combiner cancellation callback with pre_cancel_call() filter method

### PR DESCRIPTION
**Pull request by markdroth**
The call combiner ensures that only one `grpc_transport_stream_op_batch` is making its way up or down the filter stack at any given time.  Because acquiring or yielding the call combiner requires synchronization, filters that perform long-running async operations don't yield the call combiner when they start that operation and then reacquire the call combiner when the operation completes; instead, they hold the call combiner the whole time, blocking any other batches from transiting the call stack until the async operation is complete.  This is fine for most ops, but it is a problem for the `cancel_stream` op, because the whole point of cancellation is to abort any such long-running async operations and terminate the call the immediately.  So in the cancellation case, filters need a way to cancel any long-running async operation they may be doing and immediately yield the call combiner, so that the `cancel_stream` op can be sent down the filter stack.

Originally, the call combiner solved this problem by allowing a filter to register a cancellation callback, which the call combiner would invoke when the call was cancelled.  So a filter starting a long-running async operation would register a callback that, if invoked, would cancel that async operation and yeild the call combiner, thus allowing the `cancel_stream` op to start its way down the filter stack.  However, this approach has some down-sides:
- It requires synchronization to set the global call combiner callback.
- It generally requires the filter to allocate memory for the cancellation callback.
- It generally requires the callback to hold a ref on the call stack.  This has led to various bugs related to lifetime issues surrounding the cancellation callback (e.g., #16288 and #25827).
- It will make things more complicated when implementing hedging support.

This PR attempts to solve these problems by using a different approach.  Instead of having filters register a global cancellation callback with the call combiner, we add a new `pre_cancel_call()` method to the filter API.   This method works in basically the same way as the existing `start_transport_stream_op_batch()` method, except that it does *not* require holding the call combiner.  Each filter's implementation of this method must cancel any long-running async operations that might be in flight and yield the call combiner as soon as possible so that the `cancel_stream` op may be started via `start_transport_stream_op_batch()`.

This approach does require the use of additional synchronization in individual filter implementations, but in a way that should not cause any contention *except* in cancellation cases.  And in return, it eliminates the synchronization of having a global cancellation callback used by all filters, and it eliminates the need for filters to allocate memory or take a ref on the call stack for those cancellation callbacks.

### CODE REVIEW

**Comment by markdroth** [comment]
This is now ready for review.

@soheilhy, I'd appreciate it if you could give this a second review with an eye on both any obvious performance issues and on atomic-correctness.  Thanks!

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 1]
```diff
@@ -2542,7 +2554,6 @@ ClientChannel::LoadBalancedCall::LoadBalancedCall(
       call_start_time_(args.start_time),
```
Comment (soheilhy) [inline src/core/ext/filters/client_channel/client_channel.cc:2554]: Although this is a primitive, I believe it's UB since arg is `move`d on line 2425.

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 2]
```diff
@@ -1914,6 +1931,37 @@ void ClientChannel::CallData::Destroy(
...
+      // Fail pending batches on the call.
+      calld->PendingBatchesFail(elem, error,
+                                YieldCallCombinerIfPendingBatchesFound);
+      return;
+    }
+  }
+  // Not pending resolver result, so check if we have a dynamic call.
+  MutexLock lock(&calld->dynamic_call_creation_mu_);
+  if (calld->dynamic_call_ != nullptr) {
+    calld->dynamic_call_->PreCancel(error);
```
Comment (soheilhy) [inline src/core/ext/filters/client_channel/client_channel.cc:1958]: I'm not sure how long `PreCancel` would take in general, but does this call have to happen under lock? (e.g., could we just get the pointer and store a flag that the call is being run?)

File: src/core/ext/filters/client_channel/retry_filter.cc [hunk 3]
```diff
@@ -1713,6 +1689,34 @@ void RetryFilter::CallData::SetPollent(grpc_call_element* elem,
...
+      grpc_timer_cancel(&calld->retry_timer_);
+      calld->FreeAllCachedSendOpData();
+      GRPC_CALL_COMBINER_STOP(calld->call_combiner_, "Canceller");
+      return;
+    }
+  }
+  // Retry timer not pending, so propagate pre-cancellation down to LB call.
+  MutexLock lock(&calld->lb_call_pre_cancellation_mu_);
+  calld->lb_call_pre_cancelled_ = true;
+  if (calld->committed_call_ != nullptr) {
```
Comment (soheilhy) [inline src/core/ext/filters/client_channel/retry_filter.cc:1713]: nit: should there be a debug assert checking that only one of `calld->committed_call_` and `calld->call_attempt_` can be not null?

**Review by soheilhy — COMMENTED** [review]
Looks great. I have a few easy questions.

File: src/core/ext/filters/client_channel/client_channel.cc (hunk 1 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/client_channel.cc:2554]: I don't understand this comment.  The `args` that's being moved on line 2425 is a local variable in the `ClientChannel::CallData::CreateDynamicCall()` method.  The `args` that's being used here is a parameter being passed into the `ClientChannel::LoadBalancedCall::LoadBalancedCall()` ctor.  I think these are completely different objects.  Am I missing something here?

File: src/core/ext/filters/client_channel/client_channel.cc (hunk 2 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/client_channel.cc:1958]: Good point!  Done.

File: src/core/ext/filters/client_channel/retry_filter.cc (hunk 3 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/retry_filter.cc:1713]: Done.

File: src/core/ext/filters/client_channel/client_channel.cc (hunk 1 above)
Comment (soheilhy) [reply src/core/ext/filters/client_channel/client_channel.cc:2558]: Oh I'm sorry. I somehow thought this a continuation. Please disregard.

**Review by soheilhy — APPROVED** [review]

File: src/core/lib/security/transport/client_auth_filter.cc [hunk 4]
```diff
@@ -110,8 +108,51 @@ struct call_data {
...
+
+  // The error seen from a cancel_stream op.
+  grpc_error_handle cancel_error = GRPC_ERROR_NONE;
+
+  // The batch containing send_initial_metadata is stored here if we
+  // receive a pre-cancellation but have not yet seen the cancel_stream
+  // op.  This ensures that the batch is failed with the right error,
+  // which may affect the call's status if the batch also contains
+  // recv_trailing_metadata.
+  grpc_transport_stream_op_batch* send_initial_metadata_batch = nullptr;
```
Comment (yashykt) [inline src/core/lib/security/transport/client_auth_filter.cc:155]: alternatively, we could send the error with PreCancel too

**Review by yashykt — APPROVED** [review]

File: src/core/lib/security/transport/client_auth_filter.cc (hunk 4 above)
Comment (yashykt) [inline src/core/lib/security/transport/client_auth_filter.cc:155]: why not just save the error instead and cancel the batch immediately?

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 5]
```diff
@@ -340,6 +352,11 @@ class DynamicTerminationFilter::CallData {
     calld->lb_call_->StartTransportStreamOpBatch(batch);
   }
 
+  static void PreCancel(grpc_call_element* elem, grpc_error_handle error) {
+    auto* calld = static_cast<CallData*>(elem->call_data);
+    calld->lb_call_->PreCancel(error);
```
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/ext/filters/client_channel/client_channel.cc:357]: Yes.  `SetPollent()` will always have been called before this is called, because the dynamic filter stack does that on call creation:

https://github.com/grpc/grpc/blob/266d0f7c052bdb686e93525c587e2b9513b35fa7/src/core/ext/filters/client_channel/dynamic_filters.cc#L59

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 6]
```diff
@@ -2759,6 +2774,36 @@ void ClientChannel::LoadBalancedCall::StartTransportStreamOpBatch(
...
+  }
+  // Check if we have a subchannel call.
+  bool subchannel_call_exists = false;
+  {
+    MutexLock lock(&subchannel_call_creation_mu_);
+    subchannel_call_pre_cancelled_ = true;
+    if (subchannel_call_ != nullptr) subchannel_call_exists = true;
+  }
+  if (subchannel_call_exists) {
+    subchannel_call_->PreCancel(error);
```
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/ext/filters/client_channel/client_channel.cc:2801]: I believe so.  The mutex guards creation of `subchannel_call_`, and once created, it is never destroyed for the lifetime of the `LoadBalancedCall` object.

File: src/core/lib/security/transport/client_auth_filter.cc (hunk 4 above)
Comment (markdroth) [reply src/core/lib/security/transport/client_auth_filter.cc:155]: Because the error passed to `pre_cancel_call()` might not be the same as the error sent down with the `cancel_stream` op, and it's the latter value that we need to use for the `recv_trailing_metadata` batch.  (I don't remember at this point which test(s) it was that surfaced this problem, but there were test failures that required this fix.)

**Review by markdroth — COMMENTED** [review]
Thanks for the review!

File: src/core/lib/security/transport/client_auth_filter.cc (hunk 4 above)
Comment (yashykt) [reply src/core/lib/security/transport/client_auth_filter.cc:155]: Why would it not be the same? Isn't the trigger for this in call.cc? Don't we use the same error for `pre_cancel_call` and `cancel_stream`?

Comment (markdroth) [reply src/core/lib/security/transport/client_auth_filter.cc:155]: You're right, we do.  Sorry, I must have been thinking of an earlier variant of this code when I wrote that.

I looked back at which test failure prompted this change to jog my memory, and it was this one:

https://source.cloud.google.com/results/invocations/ed092efc-40e8-496d-ae9f-013e2021a8cb/targets/%2F%2Ftest%2Fcpp%2Fend2end:end2end_test@poller%3Depoll1/tests;group=ProxyEnd2end%2FProxyEnd2endTest;test=RpcDeadlineExpires%2F0;row=1

The test was seeing UNKNOWN instead of DEADLINE_EXCEEDED.  I think the problem was that the batch containing `recv_trailing_metadata` had not yet hit the filter at the point at which `pre_cancel_call()` was invoked (because `pre_cancel_call()` is not invoked from within the call combiner, so the two are not synchronized from the perspective of the filter stack).  So that batch wound up getting failed with a different error, and that's the error that dictates the status code of the call.

IIRC, the actual problem that caused that test failure was in the client_channel filter, which I fixed in 91c43c8, but then I realized that the client_auth filter would have the same problem, so I fixed it (and some other edge cases I noticed) in 7b50bdc.  It might be that none of our tests would actually trigger this case in the client_auth filter, but it still seems safer to handle it properly.

Comment (yashykt) [reply src/core/lib/security/transport/client_auth_filter.cc:155]: I don't think saving the previous batch is the right approach here. We should instead check who is responsible for invoking `recv_trailing_metadata` with the correct status/error and make sure that that error is being propagated.

**Comment by markdroth** [comment]
#26172 is a possible alternative to this.  That PR completely eliminates the `cancel_stream` op and uses the new `pre_cancel_call()` method (which is just called `cancel_call()`) to do all of the cancellation.  I'm thinking it might be a cleaner approach, but I'd welcome feedback on this.
